### DISCUSSION TITLE
How to implement sorting and tag management for large numbers of documents

### DISCUSSION THREAD

**Discussion by shuto-mito** [automerge/automerge#894] [Q&A]
I am looking to develop a local-first mobile app and am interested in this project!

However, I am wondering if it can withstand the following use cases
- Sorting 10,000 to 100,000 documents in order of edit or creation
- Managing 10,00 to 10,000 tags.

**Edit**:
> I'm not sure if this is a good estimate of the scale.
> As for scale estimates, 100,000 is a worst-case scenario.
> I'd say it's more like 20,000 to 30,000.
>
> If a user creates 10 documents per day and uses the system for 5 years, the total number of documents is 18,250. I think this is a relatively realistic scale.
>
> The UUID is 36 Bytes for a string, and 20,000 is about 720KB.
> Actually, it is less of a problem than I thought, and perhaps I am being too fastidious in my search for efficiency.
> (I don't know much about computer resources.)

Here are my ideas and their problems
### Sorting of documents
- I was thinking of having an external database (SQLite, etc.) and using it as an index for the documents, but I am not sure if this method will really work with consistency.

### Tag Management
- Given the use of an external database, we would end up implementing our own synchronization or using other libraries to synchronize additions, tag name changes, and deletions.
- Another option would be to manage all tags as a single document, but this would be inefficient as it would load all tags into memory each time.

Or is automerge not a match for a use case like mine?

Thank you very much for your time.

**Comment by jessegrosjean** [comment]
Here are some quick thoughts from a non-automerge expert (expert corrections welcome!):

1. I think storing individual document state in database is pretty standard approach.
2. I think the standard way to reference those documents is to create a well known "index" document that refers to the other documents. That way you can use automerge for all sync logic. Now will your index document be able to scale to 100,000 references, I'm not sure, but I would test that approach before adding a separate sync mechanism.
3. For tags, instead of managing them using separate sync mechanism it might work to tag documents directly using automerge state and then just extract those tags into your database for searching when you sync documents. I think this design might be simpler assuming individual documents won't have a huge amount of tags. If an individual document can have 100,000 tags then a separate tagging mechanism is likely needed.

**Edit** If you haven't found it yet you should also look at "automerge-repo". It's already doing the document management into database part. Will it scale to 100,000 documents, I'm not sure, but I expect this is the sort of thing they would like it to be able to do eventually.

**Reply by shuto-mito** [reply]
Thanks for the answers! Thanks to you, I now have a better understanding of a few questions.

1. regarding the approach of storing document state in the database, I see that automerge-repo is a library for that purpose.
2. regarding indexed documents, for me, I would avoid them in a case like this. Since this application is not an MVP or prototype, I would like to avoid memory inefficient design.
3. Regarding tags, the method of tagging documents directly is problematic in some cases.
   - If a single tag is attached to a large number of documents, renaming or deleting the tag can be expensive.
   - It is also difficult to create a tag first and then attach it to a document later.
   - To solve this problem, a document is created that manages tag IDs and tag names, and, as with index documents, the more tags, the greater the waste of memory.

In my use case, I do not expect to use a large number of tags in a single document, so this is not a problem.

I think automerge has difficulties with large number of associations.

**Reply by jessegrosjean** [reply]
> Thanks for the answers! Thanks to you, I now have a better understanding of a few questions.

Glad I could help. I think your 1, 2, 3 comments generally make sense. I really have done very little with automerge, I just have followed along enough that I think I know some of the default patterns, so I figured I could provide you with some level 1 answers.

> I think automerge has difficulties with large number of associations

I think this is probably right when "large" gets really large. In particular each automerge document is an in memory data structure, not a scalable database.

**Reply by shuto-mito** [reply]
In case you are wondering, I think at least 20,000 to 30,000 is a realistic scale for a personal notes application. I presented 100,000 as a worst case scenario.

> I figured I could provide you with some level 1 answers.

Still, your answers were very helpful. Thank you so much!
